pkg/controllers/workspace_controller.go

```diff
@@ -180,6 +204,81 @@ func (c *WorkspaceReconciler) deleteWorkspace(ctx context.Context, wObj *kaitov1
 	return c.garbageCollectWorkspace(ctx, wObj)
 }
 
+func (c *WorkspaceReconciler) updateControllerRevision(ctx context.Context, wObj *kaitov1alpha1.Workspace) error {
+	currentHash := computeHash(wObj)
+	annotations := wObj.GetAnnotations()
+
+	latestHash, exists := annotations[WorkspaceRevisionAnnotation]
+	if exists && latestHash == currentHash {
+		return nil
+	}
+
+	data := map[string]string{"hash": currentHash}
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("failed to marshal revision data: %w", err)
+	}
+
+	revisions := &appsv1.ControllerRevisionList{}
+	if err := c.List(ctx, revisions, client.InNamespace(wObj.Namespace), client.MatchingLabels{WorkspaceNameLabel: wObj.Name}); err != nil {
+		return fmt.Errorf("failed to list revisions: %w", err)
+	}
+	sort.Slice(revisions.Items, func(i, j int) bool {
+		return revisions.Items[i].Revision < revisions.Items[j].Revision
+	})
+
+	revisionNum := int64(1)
+
+	if exists {
```

Comment:
This code is dangerous. You can have a case where the annotation exists, but all controller revisions are deleted by user. Then this code hits a crash because len(revisions.Items) == 0.  You should always check the sorted list and pick the largest revision number from there.

Reply:
To add, when you first make controller there would be zero controller revision objects. So this line below will panic - out of bounds.

Reply:
> To add, when you first make controller there would be zero controller revision objects. So this line below will panic - out of bounds. tricky bug I didn't spot it at first either :)

First have controller revision? In this case I think it should not, since it will not go into  `if exists`